### PR TITLE
[ACS][CallAutomatoin] Easily Handling EventHandler 

### PR DESCRIPTION
**Pull request by minwoolee-msft**
ACS Call Automation is looking to add a new Event Processor to provide an easier way of handling event.

**Allow Event Processor to process event**
```csharp
// wait for event...
_eventHandler = _callAutomationClient.GetEventProcessor();
_eventHandler.ProcessEvents(cloudEvents);
```

**Waiting on Event with result model (easiest way)**
```csharp
AddParticipantsResult result= callConnection.AddParticipants(...);
AddParticipantsEventResult eventResult = await result.WaitForEvent(); // will wait until event arrives, or times out
AddParticipantsSucceeded myEvent = eventResult.SuccessEvent; // This is AddParticipantSucceeded event!
```

**Waiting on any event with predicate (more advanced)**
```csharp
// will wait until event arrives, or times out for Call disconnect
CallDisconnected disconnectedEvent = (CallDisconnected)await _eventHandler.WaitForEvent(predicate 
=> predicate.CallConnectionId == myConnectionId 
&& predicate.GetType() == typeof(CallDisconnected)); 
```

**Waiting any event any time, then run delegates on match (great for non-sequential events, like ParticipantsUpdated or CallDisconnect)**
```csharp
// whenever participantsUpdated event comes, print out # of people in the call!
_eventHandler.AttachOngoingEventProcessor<ParticipantsUpdated>(myCallConnectionId,
e => _logger.LogInformation($"Current # of People in the call: {e.Participants.Count}"));
```

**Significant changes that need attention:**
1. All operation generate operation context if not provided, SDK will generate one for them and send in the request. This will not impact any existing behaviors- but this is required to have eventProcessor work properly.
2. Some methods that did not return result, will now return Result instead of Void (such as, PlayResult). This is required for it to work with Result.WaitForEvent() Pattern. This will not impact any existing behaviors.

### CODE REVIEW

File: sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/EventAwaiterOngoing.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+using System;
+
+namespace Azure.Communication.CallAutomation
+{
+    internal class EventAwaiterOngoing<TEvent> where TEvent : CallAutomationEventBase
+    {
+        private string _callConnectionId;
+        private Action<TEvent> _eventHandler;
+
+        internal Action<object, CallAutomationEventArgs> OnEventRecieved => OnEventsReceived;
```
Comment (DominikMe) [inline sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/EventAwaiterOngoing.cs:13]: typo:
```suggestion
        internal Action<object, CallAutomationEventArgs> OnEventReceived => OnEventsReceived;
```
(and update references)

File: sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventHandler.cs [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            ProcessEvents(recievedEvent);
+        }
+
+        /// <summary>
+        /// Process incoming events. Pass incoming events to get it processed to have other method like WaitForEvent to function.
+        /// </summary>
+        /// <param name="events">Incoming CallAutomationEventBase object.</param>
+        public void ProcessEvents(IEnumerable<CallAutomationEventBase> events)
+        {
+            CallAutomationEventBase recievedEvent = events.FirstOrDefault();
```
Comment (DominikMe) [inline sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventHandler.cs:47]: typo
```suggestion
            CallAutomationEventBase receivedEvent = events.FirstOrDefault();
```

File: sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/EventBacklog.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+                return _eventBacklog.TryAdd(backlogEventId, (eventsToBeSaved, expiringTimer));
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        internal bool GetAndRemoveEvent(IEnumerable<Type> eventTypes, string callConnectionId, string operationContext, out KeyValuePair<string, CallAutomationEventBase> matchingEvent)
```
Comment (petrsvihlik) [inline sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/EventBacklog.cs:51]: This is a typical example for Try Pattern. The method should be prefixed with `Try`

File: sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/EventBacklog.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            _eventBacklog = new ConcurrentDictionary<string, (CallAutomationEventBase, Timer)>();
+        }
+
+        /// <summary>
+        /// Adds event to EventsBacklog to be retrieved.
+        /// </summary>
+        /// <param name="backlogEventId">Internally used id for tracking the saved event.</param>
+        /// <param name="eventsToBeSaved">Incoming Event to be saved.</param>
+        /// <returns></returns>
+        internal bool AddEvent(string backlogEventId, CallAutomationEventBase eventsToBeSaved)
```
Comment (petrsvihlik) [inline sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/EventBacklog.cs:36]: This is a typical example for Try Pattern. The method should be prefixed with `Try`

File: sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/EventBacklog.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Remove the event by Id.
+        /// </summary>
+        /// <param name="internalEventId">Key of the event in events backlog.</param>
+        /// <returns></returns>
+        internal bool RemoveEvent(string internalEventId) => _eventBacklog.TryRemove(internalEventId, out _);
```
Comment (petrsvihlik) [inline sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/EventBacklog.cs:77]: TryRemoveEvent

**Review by petrsvihlik — CHANGES_REQUESTED** [review]

File: sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventHandler.cs [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+            ProcessEvents(receivedEvent);
+        }
+
+        /// <summary>
+        /// Process incoming events. Pass incoming events to get it processed to have other method like WaitForEvent to function.
+        /// </summary>
+        /// <param name="events">Incoming CallAutomationEventBase object.</param>
+        public void ProcessEvents(IEnumerable<CallAutomationEventBase> events)
+        {
+            CallAutomationEventBase receivedEvent = events.FirstOrDefault();
```
[the remark this answers is not in the document]
Comment (minwoolee-msft) [reply sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventHandler.cs:47]: The reason is that events received will always be (as of now) single event.
I think when I was doing development, I assumed there will be no changes that will register more than 1 (as that is the case right now)
I can update this to work with multiple events passed, for future proof.

File: sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventArgs.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+using System;
+
+namespace Azure.Communication.CallAutomation
+{
+    internal class CallAutomationEventArgs : EventArgs
+    {
+        internal string eventArgsId { get; set; }
+
+        internal CallAutomationEventBase callAutomationEvent { get; set; }
```
[the remark this answers is not in the document]
Comment (minwoolee-msft) [reply sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventArgs.cs:12]: Thanks - Will Update.

File: sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/EventAwaiter.cs [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        /// </summary>
+        private void TimerProc(object state)
+        {
+            taskSource.TrySetException(new[] { new TimeoutException() });
+        }
+
+        public void Dispose()
+        {
+            _expiringTimer.Dispose();
+        }
```
[the remark this answers is not in the document]
Comment (minwoolee-msft) [reply sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/EventAwaiter.cs:67]: Thank you - will read and follow the guideline's suggestion.

File: sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventHandler.cs [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+using Azure.Messaging;
+using System.Linq;
+using System.Collections.Concurrent;
+
+namespace Azure.Communication.CallAutomation
+{
+    /// <summary>
+    /// Call Automation's EventHandler for incoming events for ease of use.
+    /// </summary>
+    public class CallAutomationEventHandler
```
[the remark this answers is not in the document]
Comment (minwoolee-msft) [reply sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventHandler.cs]: Updated to EventProcessor

File: sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventHandler.cs [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+        /// <param name="callConnectionId">CallConnectionId of the call.</param>
+        /// <param name="operationContext">(Optional) Optional operationContext of the method call.</param>
+        /// <returns>Returns CallAutomationEvent once matching event arrives.</returns>
+        public async Task<CallAutomationEventBase> WaitForEvent<T1, T2, T3, T4, T5>(string callConnectionId, string operationContext = null)
+            where T1 : CallAutomationEventBase
+            where T2 : CallAutomationEventBase
+            where T3 : CallAutomationEventBase
+            where T4 : CallAutomationEventBase
+            where T5 : CallAutomationEventBase
+            => await WaitForEvent(new List<Type> { typeof(T1), typeof(T2), typeof(T3), typeof(T4), typeof(T5) }, callConnectionId, operationContext).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (minwoolee-msft) [reply sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventHandler.cs]: Added a lot more testing for each cases

File: sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventHandler.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (petrsvihlik) [reply sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventHandler.cs]: thanks for explaining!

>I can update this to work with multiple events passed, for future proof.

either that or at least leave a comment for future generations reading the code :) it's not intuitive at the first glance.

File: sdk/communication/Azure.Communication.CallAutomation/src/EventProcessor/EventProcessor.cs [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+            RemoveFromOngoingEvent(callConnectionId, typeof(TEvent));
+        }
+
+        /// <summary>
+        /// Wait for any matching incoming event for the call. Returns the event once it arrives in ProcessEvent method.
+        /// </summary>
+        /// <param name="predicate">Predicate for waiting on event</param>
+        /// <param name="eventTimeout">Timeout for this waiting on event</param>
+        /// <returns>Returns CallAutomationEvent once matching event arrives.</returns>
+        public async Task<CallAutomationEventBase> WaitForEvent(Func<CallAutomationEventBase, bool> predicate, TimeSpan eventTimeout = default)
```
[the remark this answers is not in the document]
Comment (petrsvihlik) [reply sdk/communication/Azure.Communication.CallAutomation/src/EventProcessor/EventProcessor.cs:114]: do we want/need to expose it publicly? are we expecting customers to come up with more complex/granular predicates than those defined by us?

if so, maybe it'd make sense to distinguish the method even further from Attach/Detach methods by naming it `WaitForSingleEvent`. 

and/or provide some overload(s) for common operations... like `WaitForSingleEvent<TEvent>(string connectionId = null, string operationContext = null, Func<CallAutomationEventBase, bool> predicate = null)`

**Review by petrsvihlik — COMMENTED** [review]
mostly nits

File: sdk/communication/Azure.Communication.CallAutomation/src/EventProcessor/EventResult/StartRecognizingEventResult.cs [hunk 12]
```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Azure.Communication.CallAutomation
+{
+    /// <summary>
+    /// <summary> StartRecognizingEventResult is returned from WaitForEvent of StartRecognizingResult. </summary>
```
[the remark this answers is not in the document]
Comment (minwoolee-msft) [reply sdk/communication/Azure.Communication.CallAutomation/src/EventProcessor/EventResult/StartRecognizingEventResult.cs:7]: Thanks - will update all of them.

File: sdk/communication/Azure.Communication.CallAutomation/src/EventProcessor/EventAwaiter.cs [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+            taskSource = new TaskCompletionSource<EventProcessorArgs>(TaskCreationOptions.RunContinuationsAsynchronously);
+        }
+
+        private void OnEventsReceived(object sender, EventProcessorArgs e)
+        {
+            // see if events sent matches filter set on constructor.
+            if (_predicate(e.CallAutomationEvent))
+            {
+                // dispose expiring timer, as we don't want timer activating.
```
[the remark this answers is not in the document]
Comment (minwoolee-msft) [reply sdk/communication/Azure.Communication.CallAutomation/src/EventProcessor/EventAwaiter.cs:42]: Yep - will go through the items

File: sdk/communication/Azure.Communication.CallAutomation/src/EventProcessor/EventProcessor.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (minwoolee-msft) [reply sdk/communication/Azure.Communication.CallAutomation/src/EventProcessor/EventProcessor.cs:114]: Thank you for the suggestion - I agree that waiting for single event with old way of doing things (with generic event <TEvent>) is very easy to use, and I can see this is 90% of the time how devs would use it. I will have them added as overload.

**Review by petrsvihlik — APPROVED** [review]

**Pull request merged**